File: auxiliary/mylinearsolve.m
function [alpha, beta] = mylinearsolve(X, b, n)
    % For large scale
    zeta = pcg(@myresidual, b, 1e-8, 100);
    function Ax =  myresidual(x)
        xtop = x(1:n,1);
        xbottom = x(n+1:end,1);
        Axtop = xtop + X*xbottom;
        Axbottom = X'*xtop + xbottom;
        Ax = [Axtop; Axbottom];
    end
    alpha = zeta(1:n, 1);
    beta = zeta(n+1:end, 1);
end  

File: myImplementation/mylinearsolve.py
import numpy as np
from scipy.sparse.linalg import cg
from scipy.sparse.linalg import LinearOperator

def mylinearsolve(X, b, n):
    def myresidual(x):
        xtop = x[:n]
        xbottom = x[n:]
        Axtop = xtop + (X@xbottom)
        Axbottom = (X.conj().T @ xtop) + xbottom
        Ax = np.concatenate((Axtop, Axbottom))
        return Ax
    myresid = LinearOperator((2*n, 2*n), matvec=myresidual)

    zeta, _ = cg(myresid, b, tol=1e-8, maxiter=100)
    alpha = zeta[:n]
    beta = zeta[n:]
    alpha, beta = alpha.reshape((n, -1)), beta.reshape((n, -1))

    return alpha, beta